android/java/org/chromium/chrome/browser/crypto_wallet/util/Utils.java

```diff
@@ -941,39 +944,55 @@ public static String getAccountName(AccountInfo[] accounts, String address) {
         return stripAccountAddress(address);
     }
 
-    public static void openTransaction(
-            TransactionInfo txInfo, JsonRpcService jsonRpcService, Activity activity) {
-        if (txInfo == null || txInfo.txHash == null || txInfo.txHash.isEmpty()) {
-            return;
-        }
-        assert jsonRpcService != null;
-        jsonRpcService.getChainId(chainId -> {
-            jsonRpcService.getAllNetworks(networks -> {
-                for (EthereumChain network : networks) {
-                    if (!chainId.equals(network.chainId)) {
-                        continue;
-                    }
-                    String blockExplorerUrl = Arrays.toString(network.blockExplorerUrls);
-                    if (blockExplorerUrl.length() > 2) {
-                        blockExplorerUrl =
-                                blockExplorerUrl.substring(1, blockExplorerUrl.length() - 1)
-                                + "/tx/" + txInfo.txHash;
-
-                        TabUtils.openUrlInNewTab(false, blockExplorerUrl);
-                        TabUtils.bringChromeTabbedActivityToTheTop(activity);
-                        break;
+    public static void openTransaction(TransactionInfo txInfo, JsonRpcService jsonRpcService,
+            AppCompatActivity activity, String accountName) {
+        if (txInfo.txStatus == TransactionStatus.UNAPPROVED) {
+            if (activity instanceof ApprovedTxObserver) {
+                showApproveDialog(txInfo, accountName, activity, ((ApprovedTxObserver) activity));
+                return;
+            }
+            throw new RuntimeException("Activity must implement ApprovedTxObserver");
+        } else {
+            if (txInfo == null || txInfo.txHash == null) {
```

Comment:
Need to move `if (txInfo == null)` as a top check as it doesn't make sense that we check it after trying to access `txInfo.txStatus`

Reply:
I assume a transaction can only be opened if it was there (not null) in the list so I didn't do it for unapproved case though can add this check if it is preferred (it's there because I kept the old code as it is).

Reply:
let's put `assert txInfo != null;` and remove `txInfo == null`. The assert has to be done here
```
public static void openTransaction(TransactionInfo txInfo, JsonRpcService jsonRpcService,
            AppCompatActivity activity, AccountInfo[] accountInfos)
```